src/parser/desugar.ml

```diff
@@ -388,13 +362,36 @@ let rec comp ~yield (env : Value.env) bound (c',loc) =
       let c = comp ~yield env bound c in
       Syntax.Refl c, loc
 
+    | Input.Signature (s,cs) ->
+      begin match Value.get_signature s env with
+        | Some s_def ->
+          let rec fold bound xcs def cs = match def,cs with
+            | [], [] ->
+              Syntax.Signature (s,List.rev xcs), loc
+            | (l,_,_)::def, (l',mx,mc)::cs ->
+              if Name.eq_ident l l'
+              then
+                let x = match mx with | Some x -> x | None -> l in
+                let mc = match mc with
+                  | Some c -> Some (comp ~yield env bound c)
+                  | None -> None
+                in
+                let bound = add_bound x bound in
+                fold bound ((x,mc)::xcs) def cs
+              else Error.syntax ~loc "Bad signature constraint: labels do not match.\nConstraints must be fully described."
+            | _::_,[] | [],_::_ -> Error.syntax ~loc "Bad signature constraint: lengths do not match.\nConstraints must be fully described."
+          in
+          fold bound [] s_def cs
+        | None -> Error.syntax ~loc "unknown signature %t" (Name.print_ident s)
+      end
+
     | Input.Structure lycs ->
        let s = find_signature ~loc env (List.map (fun (l,_,_)->l) lycs) in
```

Comment:
How is this supposed to work? Desugar cannot possibly figure out the signature by looking at the structure since any number of fields could be missing. Even if `find_signature` finds one that matches the labels perfectly, we could still be missing some labels. Is this a TODO?